// @commerce-apps/core/dist/index.js
"use strict";
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StaticClient = void 0;
var client_1 = require("./base/client");
Object.defineProperty(exports, "BaseClient", { enumerable: true, get: function () { return client_1.BaseClient; } });
Object.defineProperty(exports, "ClientConfig", { enumerable: true, get: function () { return client_1.ClientConfig; } });
Object.defineProperty(exports, "Response", { enumerable: true, get: function () { return client_1.Response; } });
Object.defineProperty(exports, "ResponseError", { enumerable: true, get: function () { return client_1.ResponseError; } });
const staticClient_1 = require("./base/staticClient");
var staticClient_2 = require("./base/staticClient");
Object.defineProperty(exports, "getObjectFromResponse", { enumerable: true, get: function () { return staticClient_2.getObjectFromResponse; } });
Object.defineProperty(exports, "getHeaders", { enumerable: true, get: function () { return staticClient_2.getHeaders; } });
Object.defineProperty(exports, "mergeHeaders", { enumerable: true, get: function () { return staticClient_2.mergeHeaders; } });
exports.StaticClient = {
    get: staticClient_1._get,
    delete: staticClient_1._delete,
    patch: staticClient_1._patch,
    post: staticClient_1._post,
    put: staticClient_1._put,
};
var authHelper_1 = require("./base/authHelper");
Object.defineProperty(exports, "ShopperToken", { enumerable: true, get: function () { return authHelper_1.ShopperToken; } });
Object.defineProperty(exports, "stripBearer", { enumerable: true, get: function () { return authHelper_1.stripBearer; } });
var commonParameters_1 = require("./base/commonParameters");
Object.defineProperty(exports, "commonParameterPositions", { enumerable: true, get: function () { return commonParameters_1.commonParameterPositions; } });
var cacheManagerKeyv_1 = require("./base/cacheManagerKeyv");
Object.defineProperty(exports, "CacheManagerKeyv", { enumerable: true, get: function () { return cacheManagerKeyv_1.CacheManagerKeyv; } });
var cacheManagerRedis_1 = require("./base/cacheManagerRedis");
Object.defineProperty(exports, "CacheManagerRedis", { enumerable: true, get: function () { return cacheManagerRedis_1.CacheManagerRedis; } });
var sdkLogger_1 = require("./base/sdkLogger");
Object.defineProperty(exports, "COMMERCE_SDK_LOGGER_KEY", { enumerable: true, get: function () { return sdkLogger_1.COMMERCE_SDK_LOGGER_KEY; } });
Object.defineProperty(exports, "sdkLogger", { enumerable: true, get: function () { return sdkLogger_1.sdkLogger; } });
